spec/derived_build_code_formatter_spec.rb

```diff
@@ -81,6 +74,125 @@
+      end
+
+      it 'uses mixed digit counts' do
+        version = Fastlane::Models::AppVersion.new(1, 23, 45, 678)
+        formatter = described_class.new(prefix: '', major_digits: 1, minor_digits: 2, patch_digits: 2, build_digits: 3)
+        build_code_string = formatter.build_code(version: version)
+        # 1(1 digit) + 23(2 digits) + 45(2 digits) + 678(3 digits) = "12345678"
+        expect(build_code_string.to_s).to eq('12345678')
+      end
+
+      it 'handles maximum values within digit limits' do
+        version = Fastlane::Models::AppVersion.new(99, 99, 99, 99)
+        formatter = described_class.new(prefix: '1', major_digits: 2, minor_digits: 2, patch_digits: 2, build_digits: 2)
+        build_code_string = formatter.build_code(version: version)
+        expect(build_code_string.to_s).to eq('199999999')
+      end
+    end
+
+    context 'with empty prefix and custom digits' do
+      it 'trims leading zeros correctly with 1-digit major' do
+        version = Fastlane::Models::AppVersion.new(5, 12, 34, 56)
+        formatter = described_class.new(prefix: '', major_digits: 1, minor_digits: 2, patch_digits: 2, build_digits: 2)
+        build_code_string = formatter.build_code(version: version)
+        expect(build_code_string.to_s).to eq('5123456')
+      end
+
+      it 'trims leading zeros correctly with larger number of major digits' do
+        version = Fastlane::Models::AppVersion.new(7, 8, 9, 10)
+        formatter = described_class.new(prefix: '', major_digits: 2, minor_digits: 2, patch_digits: 2, build_digits: 2)
+        build_code_string = formatter.build_code(version: version)
+        expect(build_code_string.to_s).to eq('7080910')
+      end
+
+      it 'handles edge case where all components start with zeros' do
+        version = Fastlane::Models::AppVersion.new(0, 1, 2, 3)
+        formatter = described_class.new(prefix: '', major_digits: 2, minor_digits: 2, patch_digits: 2, build_digits: 2)
+        build_code_string = formatter.build_code(version: version)
+        # ''(empty prefix) + '00'(2 digits) + '01'(2 digits) + '02'(2 digits) + '03'(2 digits) = "00010203", trimmed to "10203"
+        expect(build_code_string.to_s).to eq('10203')
+      end
+    end
+
+    context 'with backward compatibility (default 2 digits)' do
+      it 'maintains existing behavior when no digit parameters specified' do
+        version = Fastlane::Models::AppVersion.new(1, 2, 3, 4)
+        formatter_old = described_class.new(prefix: '1')
+        formatter_new = described_class.new(prefix: '1', major_digits: 2, minor_digits: 2, patch_digits: 2, build_digits: 2)
+
+        expect(formatter_old.build_code(version: version)).to eq(formatter_new.build_code(version: version))
+      end
+    end
+  end
+
+  describe 'digit count validation' do
+    # Generate all combinations of [1,2,3] for each of the 4 digit count parameters
+    all_digit_combinations = Array.new(4, [1, 2, 3]).reduce(&:product).map(&:flatten)
```

Comment:
💭 I'm surprised that `rubocop` doesn't complain about you using a local variable here instead of using RSpec's `let(:all_digit_combinations) { … }` syntax (which I believed was more idiomatic to Rspec)…

I don't mind either approach in practice (should do the same in the end, maybe except some memoizing of the value when using `let`?). I just wouldn't want a `rubocop` warning to appear because of this?